src/Diagnostics/DiagnosticsMachine/group_gen.jl

```diff
@@ -0,0 +1,602 @@
+            # Adjust `level` on the sphere.
+            level_val = dims["level"]
+            dims["level"] = (
+                level_val[1] .- FT(planet_radius(Settings.param_set)),
+                level_val[2],
+            )
+        end
+        dims
+    end
+end
+
+get_dimnames(::NoInterpolation, cfg, dvtype) =
+    DiagnosticsMachine.dv_dg_dimnames(cfg, dvtype)
+get_dimnames(::InterpolationType, cfg, dvtype) =
+    :(tuple(collect(keys(dims))...))
+
+# Generate the `vars` dictionary for `Writers.init_data`.
+function generate_init_vars(intrp, cfg, dvtype_dvars_map)
+    varslst = Any[]
+    for (dvtype, dvlst) in dvtype_dvars_map
+        for dvar in dvlst
+            lhs = :($(DiagnosticsMachine.dv_name(cfg, dvar)))
+            dimnames = get_dimnames(intrp, cfg, dvtype)
+            rhs = :($dimnames, FT, $(DiagnosticsMachine.dv_attrib(cfg, dvar)))
+            push!(varslst, :($lhs => $rhs))
+        end
+    end
+
+    quote
+        # TODO: add code to filter this based on what's actually in `bl`.
+        OrderedDict($(Expr(:tuple, varslst...))...)
+    end
+end
+
+# Generate `Diagnostics.$(name)_init(...)` which will initialize the
+# `DiagnosticsGroup` when called.
+function generate_init_fun(
+    name,
+    config_type,
+    params_type,
+    init_fun,
+    interpolate,
+    dvtype_dvars_map,
+)
+    init_name = Symbol(name, "_init")
+    cfg = getfield(ConfigTypes, config_type)
+    intrp = getfield(@__MODULE__, interpolate)
+    quote
+        function $init_name(dgngrp, curr_time)
+            $(generate_common_defs())
+
+            $(init_fun)(dgngrp, curr_time)
+
+            # temporarily if dgngrp.onetime
+            DiagnosticsMachine.collect_onetime(mpicomm, dg, Q)
+            # temporarily end
+
+            if mpirank == 0
+                dims = $(generate_init_dims(intrp(), cfg(), dvtype_dvars_map))
+                vars = $(generate_init_vars(intrp(), cfg(), dvtype_dvars_map))
+
+                # create the output file
+                dprefix = @sprintf("%s_%s", dgngrp.out_prefix, dgngrp.name,)
+                dfilename = joinpath(Settings.output_dir, dprefix)
+                init_data(dgngrp.writer, dfilename, dims, vars)
+            end
+
+            return nothing
+        end
+    end
+end
+
+# Generate code snippet for copying arrays to the CPU if needed. Ideally,
+# this will be removed when diagnostics are made to run on GPU.
+function generate_array_copies()
+    quote
+        # get needed arrays onto the CPU
+        if array_device(Q) isa CPU
+            state_data = Q.realdata
+            gradflux_data = dg.state_gradient_flux.realdata
+            aux_data = dg.state_auxiliary.realdata
+            vgeo = grid.vgeo
+        else
+            state_data = Array(Q.realdata)
+            gradflux_data = Array(dg.state_gradient_flux.realdata)
+            aux_data = Array(dg.state_auxiliary.realdata)
+            vgeo = Array(grid.vgeo)
+        end
+    end
+end
```

Comment:
This seems a bit dubious to me. Is this meant to be a closure?

Reply:
Not sure what you mean? It's an expression that's embedded into the generated function [here](https://github.com/CliMA/ClimateMachine.jl/blob/kp/diagnostics-machine/src/Diagnostics/DiagnosticsMachine/group_gen.jl#L492). It's pretty much the same thing that exists in current diagnostics groups (e.g. [here](https://github.com/CliMA/ClimateMachine.jl/blob/kp/diagnostics-machine/src/Diagnostics/atmos_les_default.jl#L588-L605) or [here](https://github.com/CliMA/ClimateMachine.jl/blob/kp/diagnostics-machine/src/Diagnostics/atmos_gcm_default.jl#L327-L336)).

Reply:
Ah, I think I missed the quote. I see.